elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/engine/SQLQueryEngine.java

```diff
@@ -0,0 +1,494 @@
+     * @param query The client query
+     * @return A SQL fragment to use in the SELECT .. statement.
+     */
+    private String extractProjection(Query query) {
+        List<String> metricProjections = query.getMetrics().entrySet().stream()
+                .map((entry) -> {
+                    Metric metric = entry.getKey();
+                    Class<? extends Aggregation> agg = entry.getValue();
+                    return metric.getMetricExpression(Optional.of(agg)) + " AS " + metric.getName();
+                })
+                .collect(Collectors.toList());
+
+        List<String> dimensionProjections = query.getDimensions().stream()
+                .map(Dimension::getName)
+                .map((name) -> getColumnName(query.getSchema().getEntityClass(), name))
+                .collect(Collectors.toList());
+
+        String projectionClause = metricProjections.stream()
+                .collect(Collectors.joining(","));
+
+        if (!dimensionProjections.isEmpty()) {
+            projectionClause = projectionClause + "," + dimensionProjections.stream()
+                    .map((name) -> query.getSchema().getAlias() + "." + name)
+                    .collect(Collectors.joining(","));
+        }
+
+        return projectionClause;
+    }
+
+    /**
+     * Extracts a GROUP BY SQL clause.
+     * @param query A client query
+     * @return The SQL GROUP BY clause
+     */
+    private String extractGroupBy(Query query) {
+        List<String> dimensionProjections = query.getDimensions().stream()
+                .map(Dimension::getName)
+                .map((name) -> getColumnName(query.getSchema().getEntityClass(), name))
+                .collect(Collectors.toList());
+
+        return "GROUP BY " + dimensionProjections.stream()
+                    .map((name) -> query.getSchema().getAlias() + "." + name)
+                    .collect(Collectors.joining(","));
+
+    }
+
+    /**
+     * Converts a filter predicate into a SQL WHERE clause column reference.
+     * @param predicate The predicate to convert
+     * @return A SQL fragment that references a database column
+     */
+    private String generateWhereClauseColumnReference(FilterPredicate predicate) {
+        Path.PathElement last = predicate.getPath().lastElement().get();
+        Class<?> lastClass = last.getType();
+
+        return FilterPredicate.getTypeAlias(lastClass) + "." + getColumnName(lastClass, last.getFieldName());
+    }
+
+    /**
+     * Converts a filter predicate into a SQL HAVING clause column reference.
+     * @param predicate The predicate to convert
+     * @return A SQL fragment that references a database column
+     */
+    private String generateHavingClauseColumnReference(FilterPredicate predicate, Query query) {
+        Path.PathElement last = predicate.getPath().lastElement().get();
+        Class<?> lastClass = last.getType();
+
+        if (! lastClass.equals(query.getSchema().getEntityClass())) {
```

Comment:
Extra space, Han had a similar comment earlier in this PR that should be addressed as well:
```suggestion
        if (!lastClass.equals(query.getSchema().getEntityClass())) {
```

Reply:
👍